[RHODA][Infrastructure] - Adding test and suite teardown

ikanias opened the pull request:
This PR includes:
1. Adding a new resource file: suite_and_test_teardown.resource
2. Configuring import_connect tests for suite tear down and provision tests for test tear down

Resolves Jira ticket DBAAS-763

ikanias committed:
[RHODA][Infrastructure] - Adding test and suite teardown

This PR includes:
1. Adding a new resource file: suite_and_test_teardown.resource
2. Configuring import_connect tests for suite tear down and provision tests for test tear down

Resolves Jira ticket DBAAS-763

harshkumarRH commented on resources/keywords/suite_and_test_teardown.resource:
```diff
@@ -0,0 +1,47 @@
+*** Settings ***
+Library     SeleniumLibrary
+Library     BuiltIn
+Library     String
+Library     OpenShiftLibrary
+Library     ../../utils/scripts/dbaas_utils.py
+Library     ../../utils/scripts/util.py
+Library    OpenShiftCLI
+Resource    ../object_repo/login_obj.resource
```
We probably do not need any of these

harshkumarRH commented on resources/keywords/suite_and_test_teardown.resource:
```diff
@@ -0,0 +1,47 @@
+
+Remove Existing DBaaSInstances
+    IF  '${instanceName}' != '${EMPTY}'
+       ${cmd_instances}    Set Variable    oc delete -A DBaaSInstance/${instanceName} -n ${newProject} --wait=true
```
'${instanceName}' != '${EMPTY}' Even when DBaaS Instances are provisioned from Admin view, because after provisioning them, we move to developer view, create a new project and try to deploy the DBaaSInstance on the new created Project Topology

So consequently, the final command which will get executed will look something like this:
oc delete -A DBaaSInstance/some-name -n 0000-test-project-xyz  --wait=true

However, the DBaaSInstance actually doesn't belong to the 0000-test-project-xyz namespace, it belongs to the redhat-dbaas-operator namespace, so we will see an error saying resource not found and the DBaaSInstance will not get deleted.

harshkumarRH commented on resources/keywords/suite_and_test_teardown.resource:
```diff
@@ -0,0 +1,47 @@
+
+ Remove Existing DBaaSProviderAccounts
+    IF  '${provaccname}' != '${EMPTY}'
+       ${cmd_provider_account} =   Set Variable    oc delete -A DBaaSInventory/${provaccname} -n ${newProject} --wait=true
```
Similar issue like what was mentioned for DBaaSInstance removal
We need to do away with -A and -n ${newProject}
${newProject} will point to a test project we just created for deploying dbInstances and retrieving data
Whereas the DBaaSInventory(Provider Accounts) always belong to redhat-dbaas-opearator namespace, at least as of now

harshkumarRH commented on resources/keywords/suite_and_test_teardown.resource:
```diff
@@ -0,0 +1,47 @@
+
+Remove Existing DBaaSNamespaces
+    IF  '${newProject}' != '${EMPTY}'
+       ${cmd_namespaces}   Set Variable    oc delete -A Project/${newProject} --wait=true
```
Deletion of namespace sometimes get stuck, the basic reason behind this is - some of the resources within the namespace are yet to be removed, and hence the spec finalizer gets stuck waiting for those resources to be removed first.
so to avoid any kind of huge waiting time, we should also include --timeout=Xs with --wait=true

harshkumarRH commented:
@ikanias 
Here are my queries / concerns -
- Instead of adding the line "Resource            ../resources/keywords/suite_and_test_teardown.resource" to each of the test suites, we can add "Resource    ../keywords/login.resource" to suite_test_teardown.resource
- We cannot use "-A" in of the delete commands as we are targetting deletion of specific CRDs and -A will not work when resource name is also mentioned.
- Logging out of OC CLI is not required as part of Test Case Teardown, as OC Login sessions are not terminal exclusive, they are valid for the machine, so we need not logout. Also a new OC login overrides the previous one. By doing this we can decrease the time it takes for tear-down. Consequently Login to Openshift CLI will also not be needed under Test Suite Teardown
- Deletion of DBaaSInstances should be part of Test Suite teardown instead of Test Case Teardown, because if we remove the newly created DBaaSInstance at Test Case level, we won't be able to leverage it later on if we add further test cases to Dev_provision or Admin Provision Test Suites. By adding it to Test Case teardown, we are also trying to delete DBaaSInstances that we haven't created from RHODA side in Import_Connect Test Suites. Currently this may not be a big concern, but we should align this to avoid any complications that may arise in the future
- Remove Existing DBaaSNamespaces - Deletion of DBaaSNamespaces should be moved to Test Case level, as we create a Test Project with almost every test case we run, if we remove it at suite level, we will only be removing the test project which was created by the last test case all the test projects by previous test cases will remain intact

PLMKYT

ikanias commented:
Created a new PR with all the relevant fixes. Closing this one.

ikanias committed:
[RHODA][Infrastructure] - Suite and Test tear down final version

This PR contains all the fixes for all the comments added to previous PR: [RHODA][Infrastructure] - Adding test and suite teardown #54
The previous PR was closed.

Resolved Jira ticket DBAAS-763

mrrajan commented on resources/keywords/suite_and_test_teardown.resource:
```diff
@@ -0,0 +1,57 @@
+        ${stdout} =    Execute Command    ${cmd_provider_account}    True
+    END
+
+Force Delete Terminating Namespaces
```
IMHO force deleting namespace which are stuck in termination can be done manually and not in favor of keeping it on part of our framework. 
But this will not be stopping our PR merge and we discuss this later.

ikanias replied:
OK sure.

mrrajan commented:
[run_results.zip](https://github.com/RHODA-lab/rhoda-ci/files/9367308/rhoda-666-2022-08-17-2421.zip)

I have concerns on tearing down the Project on Test case level as this will impact the execution of subsequent test cases. We could either remove the suite/test variable while removing resources or avoid deleting resources on test case level.
Example --> Scenario: Provision CockroachDB Database Instance from Developers View
Please LMKWYT?

mrrajan commented on resources/keywords/provision_dbinstance.resource:
```diff
@@ -70,6 +70,7 @@ DBSC Instance Provisioned Successfully
     Element Should Be Visible    ${txtCreationInfo}
     Element Should Be Visible    ${btnViewDBInst}
     Click Button    ${btnViewDBInst}
+    Set Suite Variable ${instanceName}
```
Insufficient white space error occurs.

ikanias replied:
Fixed.

ikanias commented:
[test_report.html.tar.gz](https://github.com/RHODA-lab/rhoda-ci/files/9403132/test_report.html.tar.gz)

ikanias commented:
Hey Rajan,
If we remove this from test cases then we will not really have test case tear down.
Do you think we should leave the suite tear down only?

mrrajan reviewed (approved):
LGTM

harshkumarRH reviewed (approved):
LGTM
